### PR TITLE
LL-2279

### PR DESCRIPTION
**Pull request by LFBarreto**
(Tron): added voting flow react hooks + some tests

### CODE REVIEW

File: src/families/tron/react.js [hunk 1]
```diff
@@ -68,3 +88,165 @@ export const formatVotes = (
...
+  superRepresentatives: SuperRepresentative[],
+  votes: Vote[]
+): {
+  sr: SuperRepresentative,
+  name: ?string,
+  address: string,
+  rank: number,
+  isSR: boolean
+}[] {
+  const [initialVotes] = useState(votes.map(({ address }) => address));
```
Comment (LukeSugiura) [inline src/families/tron/react.js]: useMemo?

File: src/families/tron/react.js [hunk 2]
```diff
@@ -68,3 +88,165 @@ export const formatVotes = (
...
+        name: sr.name,
+        address: sr.address,
+        rank: rank + 1,
+        isSR: rank <= SR_THRESHOLD
+      })),
+    [superRepresentatives]
+  );
+
+  const sortedVotes = useMemo(() => {
+    return SR.filter(({ address }) => initialVotes.includes(address)).concat(
```
Comment (LukeSugiura) [inline src/families/tron/react.js]: Can omit return

File: src/families/tron/react.js (hunk 1 above)
Comment (LFBarreto) [reply src/families/tron/react.js]: not here i don't want it to recompute just initialize and that it :+1:

File: src/families/tron/react.js [hunk 3]
```diff
@@ -68,3 +88,165 @@ export const formatVotes = (
...
+    () => (search ? SR.filter(searchFilter(search)) : sortedVotes),
+    [search, SR, sortedVotes]
+  );
+
+  return sr;
+}
+
+/** Tron vote flow state reducer */
+function votesReducer(state, action: Action) {
+  const { type, address, value } = action;
```
Comment (LukeSugiura) [inline src/families/tron/react.js]: why not destructing directly on argument?

File: src/families/tron/react.js [hunk 4]
```diff
@@ -68,3 +88,165 @@ export const formatVotes = (
...
+  const sr = useMemo(
+    () => (search ? SR.filter(searchFilter(search)) : sortedVotes),
+    [search, SR, sortedVotes]
+  );
+
+  return sr;
+}
+
+/** Tron vote flow state reducer */
+function votesReducer(state, action: Action) {
```
Comment (LukeSugiura) [inline src/families/tron/react.js]: Can State type be used here?

File: src/families/tron/react.js (hunk 1 above)
Comment (LukeSugiura) [reply src/families/tron/react.js]: In that case, i would suggest useRef. useState shouldn't be used without set func.

File: src/families/tron/react.js [hunk 5]
```diff
@@ -68,3 +88,166 @@ export const formatVotes = (
...
+  onChangeVotes: (votes: Vote[]) => void,
+  tronResources: ?TronResources
+): [State, (action: Action) => void] {
+  const [state, dispatch]: [State, (action: Action) => void] = useReducer(
+    votesReducer,
+    initState(votes, tronResources)
+  );
+
+  useEffect(
+    debounce(
```
Comment (gre) [inline src/families/tron/react.js]: this makes no sense

File: src/families/tron/react.js [hunk 6]
```diff
@@ -68,3 +88,166 @@ export const formatVotes = (
...
+    votesSelected: initialVotes.length,
+    max: Math.max(0, votesAvailable - votesUsed),
+    initialVotes: votes
+  };
+}
+
+/** Hook to retrieve and update voting flow state */
+export function useVotesReducer(
+  votes: Vote[],
+  onChangeVotes: (votes: Vote[]) => void,
```
Comment (gre) [inline src/families/tron/react.js]: having both an external state and an internal state of "votes" feels like it's an antipattern; there is something to improve here. there should be no "onChangeVotes" I feel. `votes` param should be called `initialVotes` if it's truely what's happening here.

Comment (LFBarreto) [reply src/families/tron/react.js]: initially i wanted to debounce the update of the transaction votes and have an easier to use data structure for this but updating the transaction after each update doesn't seem to make the whole thing lag anymore :man_shrugging:

Comment (gre) [reply src/families/tron/react.js]: we should `useDebounce` then. because in your code, it's never calling onChangeVotes

Comment (LFBarreto) [reply src/families/tron/react.js]: the useEffect with a debounce does that, it's weird but it works i'll try with the fixed use debouncedCallback on the lld side without the debouncing here :+1:

Comment (gre) [reply src/families/tron/react.js]: ok i understand the code, but it's not optimal and it's very dangerous: if you have a lot of re-render happening, the debouncing will never be called, it's only called when there is no re-render but you don't know that

File: src/families/tron/react.js (hunk 1 above)
Comment (gre) [reply src/families/tron/react.js]: is there a reason why @LukeSugiura ? I used that pattern at a few places too and i didn't find a better one. useRef is not equivalent, is it?

**Pull request merged**